File: actionlib_pkg/src/fibonacciActionServer.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <actionlib_pkg/fibonacciAction.h>

// http://wiki.ros.org/actionlib_tutorials/Tutorials/SimpleActionServer%28ExecuteCallbackMethod%29

class FibonacciAction
{
    protected:
    ros::NodeHandle nh_;
    actionlib::SimpleActionServer<actionlib_pkg::fibonacciAction> as_; // NodeHandle instance must be created before this line. Otherwise strange error occurs.
    std::string action_name_;
    // create messages that are used to published feedback/result
    actionlib_pkg::fibonacciFeedback feedback_;
    actionlib_pkg::fibonacciResult result_;

    public:
    FibonacciAction(std::string name) :
        as_(nh_, name, boost::bind(&FibonacciAction::executeCB, this, _1), false), action_name_(name)
    {
        as_.start();
    }

    ~FibonacciAction(void){}

    void executeCB(const actionlib_pkg::fibonacciGoalConstPtr &goal){
        // helper variables
        ros::Rate r(1);
        bool success = true;

        // push_back the seeds for the fibonacci sequence
        feedback_.sequence.clear();
        feedback_.sequence.push_back(0);
        feedback_.sequence.push_back(1);

        // publish info to the console for the user
        ROS_INFO("fibonacciActionServer.cpp -> %s: Executing, creating fibonacci sequence of order %i with seeds %i, %i", action_name_.c_str(), goal->order, feedback_.sequence[0], feedback_.sequence[1]);

        // start executing the action
        for(int i=1; i<=goal->order; i++){
            // check that preempt has not been requested by the client
            if (as_.isPreemptRequested() || !ros::ok()){  // Check Cancel
                ROS_INFO("fibonacciActionServer.cpp -> %s: Preempted", action_name_.c_str());
                as_.setPreempted();  // set the action state to preempted
                success = false;
                break;
            }
            feedback_.sequence.push_back(feedback_.sequence[i] + feedback_.sequence[i-1]);
            as_.publishFeedback(feedback_); // publish the feedback
            ROS_INFO("fibonacciActionServer.cpp - executing action - publish the feedback");
            // this sleep is not necessary, the sequence is computed at 1 Hz for demonstration purposes
            r.sleep();
        }

        if(success){
            result_.sequence = feedback_.sequence;
            as_.setSucceeded(result_);  // set the action state to succeeded
            ROS_INFO("fibonacciActionServer.cpp -> %s: Succeeded", action_name_.c_str());
        }
    }
};


int main(int argc, char** argv)
{
  ros::init(argc, argv, "fibonacciActionServer");
  ROS_INFO ("fibonacciActionServer.cpp - Create node fibonacciActionServer");
  FibonacciAction fibonacci("fibonacci");
  ros::spin();

  return 0;
}

File: actionlib_pkg/scripts/fibonacciActionServer.py
#! /usr/bin/env python

import rospy
import actionlib
import actionlib_pkg.msg

# http://wiki.ros.org/actionlib_tutorials/Tutorials/Writing%20a%20Simple%20Action%20Server%20using%20the%20Execute%20Callback%20%28Python%29

class FibonacciAction(object):
    # create messages that are used to publish feedback/result
    _feedback = actionlib_pkg.msg.fibonacciFeedback()
    _result = actionlib_pkg.msg.fibonacciResult()

    def __init__(self, name):
        self._action_name = name
        self._as = actionlib.SimpleActionServer(self._action_name, actionlib_pkg.msg.fibonacciAction, execute_cb=self.execute_cb, auto_start = False)
        self._as.start()
      
    def execute_cb(self, goal):
        # helper variables
        r = rospy.Rate(1)
        success = True
        
        # append the seeds for the fibonacci sequence
        self._feedback.sequence = []
        self._feedback.sequence.append(0)
        self._feedback.sequence.append(1)
        
        # publish info to the console for the user
        rospy.loginfo('fibonacciActionServer.py -> %s: Executing, creating fibonacci sequence of order %i with seeds %i, %i' % (self._action_name, goal.order, self._feedback.sequence[0], self._feedback.sequence[1]))
        
        # start executing the action
        for i in range(1, goal.order):
            # check that preempt has not been requested by the client
            if self._as.is_preempt_requested():
                rospy.loginfo('fibonacciActionServer.py -> %s: Preempted' % self._action_name)
                self._as.set_preempted()
                success = False
                break
            self._feedback.sequence.append(self._feedback.sequence[i] + self._feedback.sequence[i-1])
            # publish the feedback
            self._as.publish_feedback(self._feedback)
            rospy.loginfo("fibonacciActionServer.py -> executing action - publish the feedback")
            # this step is not necessary, the sequence is computed at 1 Hz for demonstration purposes
            r.sleep()
          
        if success:
            self._result.sequence = self._feedback.sequence
            rospy.loginfo('fibonacciActionServer.py -> %s: Succeeded' % self._action_name)
            self._as.set_succeeded(self._result)
        
if __name__ == '__main__':
    rospy.init_node('fibonacciActionServer')
    print("fibonacciActionServer.py - Create node fibonacciActionServer")
    server = FibonacciAction("fibonacci")
    rospy.spin()